drivers/flash/flash_stm32_ospi.c

```diff
@@ -0,0 +1,1505 @@
+	LOG_DBG("Flash size %u bytes", flash_size);
+	return 0;
+}
+
+static int flash_stm32_ospi_init(const struct device *dev)
+{
+	const struct flash_stm32_ospi_config *dev_cfg = DEV_CFG(dev);
+	struct flash_stm32_ospi_data *dev_data = DEV_DATA(dev);
+	uint32_t ahb_clock_freq;
+	uint32_t prescaler = 0;
+	int ret;
+
+	/* The SPI/DTR is not a valid config of data_mode/data_rate according to the DTS */
+	if ((dev_cfg->data_mode == OSPI_SPI_MODE)
+		&& (dev_cfg->data_rate == OSPI_DTR_TRANSFER)) {
+		/* already the right config, continue */
+		LOG_ERR("OSPI mode SPI/DTR is not valid");
+		return -ENOTSUP;
+	}
+
+	/* Signals configuration */
+	ret = pinctrl_apply_state(dev_cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("OSPI pinctrl setup failed (%d)", ret);
+		return ret;
+	}
+
+	/* Initializes the independent peripherals clock */
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
+
+	/* Clock configuration */
+	if (clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+			     (clock_control_subsys_t) &dev_cfg->pclken) != 0) {
+		LOG_DBG("Could not enable OSPI clock");
+		return -EIO;
+	}
+
+	if (clock_control_get_rate(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+			(clock_control_subsys_t) &dev_cfg->pclken,
+			&ahb_clock_freq) < 0) {
+		LOG_DBG("Failed to get AHB clock frequency");
+		return -EIO;
+	}
+
+	for (; prescaler <= STM32_OSPI_CLOCK_PRESCALER_MAX; prescaler++) {
+		uint32_t clk = ahb_clock_freq / (prescaler + 1);
+
+		if (clk <= dev_cfg->max_frequency) {
+			break;
+		}
+	}
+	__ASSERT_NO_MSG(prescaler <= STM32_OSPI_CLOCK_PRESCALER_MAX);
+
+	/* Initialize OSPI HAL structure completely */
+	dev_data->hospi.Init.FifoThreshold = 4;
+	dev_data->hospi.Init.ClockPrescaler = prescaler;
+	dev_data->hospi.Init.DeviceSize = find_lsb_set(dev_cfg->flash_size);
+	dev_data->hospi.Init.DualQuad = HAL_OSPI_DUALQUAD_DISABLE;
+	dev_data->hospi.Init.ChipSelectHighTime = 2;
+	dev_data->hospi.Init.FreeRunningClock = HAL_OSPI_FREERUNCLK_DISABLE;
+	dev_data->hospi.Init.ClockMode = HAL_OSPI_CLOCK_MODE_0;
+	dev_data->hospi.Init.WrapSize = HAL_OSPI_WRAP_NOT_SUPPORTED;
+	dev_data->hospi.Init.SampleShifting = HAL_OSPI_SAMPLE_SHIFTING_NONE;
+	/* STR mode else Macronix for DTR mode */
+	if (dev_cfg->data_rate == OSPI_DTR_TRANSFER) {
+		dev_data->hospi.Init.MemoryType = HAL_OSPI_MEMTYPE_MACRONIX;
+		dev_data->hospi.Init.DelayHoldQuarterCycle = HAL_OSPI_DHQC_ENABLE;
+	} else {
+		dev_data->hospi.Init.MemoryType = HAL_OSPI_MEMTYPE_MICRON;
+		dev_data->hospi.Init.DelayHoldQuarterCycle = HAL_OSPI_DHQC_DISABLE;
+	}
+	dev_data->hospi.Init.ChipSelectBoundary = 0;
+	dev_data->hospi.Init.DelayBlockBypass = HAL_OSPI_DELAY_BLOCK_USED;
+	dev_data->hospi.Init.Refresh = 0;
+
+	if (HAL_OSPI_Init(&dev_data->hospi) != HAL_OK) {
+		LOG_ERR("OSPI Init failed");
+		return -EIO;
+	}
+
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* OCTOSPI I/O manager init Function */
+	OSPIM_CfgTypeDef ospi_mgr_cfg = {0};
+
+	__HAL_RCC_OSPIM_CLK_ENABLE();
```

Comment:
I'm surprise this is required. This should be done one `clock_control_on()` call.

Reply:
The OSPIManager is not declared as a node of the stm32U5,

Reply:
ok, thanks. There may be alternat wy to handle this via device tree, but that's ok for now.